### PR TITLE
First cut at CPU performance measurement

### PR DESCRIPTION
**Pull request by raphlinus**
This just compares against Skia, which is not especially informative.

Skia has a bug that causes over-subdivision at fine tolerances. Our code could also be more optimized; particularly low-hanging fruit includes not evaluating the Euler spiral twice for both sides of a stroke.

The SVG parser now takes into account the scale, which is especially important for the waves example.

### CODE REVIEW

**Comment by raphlinus** [comment]
I also didn't say anything about which platform I did the measurements on. This was M1 Max. It's very plausible I'll switch to AMD 3970X if I do comparisons against Nehab, as that's currently the only platform where I have it built.

I think Blend2D is probably the most important comparison, as that outputs lines.

File: paper.tex [hunk 1]
```diff
@@ -305,6 +305,23 @@ \subsection{Unrolled recursion}
...
+
+\begin{figure}
+    \includegraphics[scale=0.24]{cusp_rendering}
+    \caption{Rendering of a cubic Bézier cusp using Euler spiral segments}
+    \Description{A cubic Bézier with a cusp is approximated by Euler spiral segments with finite curvature, including one with a question mark shape. The resulting stroke outlines have the correct shape.}
+    \label{fig:cusp_rendering}
+\end{figure}
+
+As recommended in both \cite{Nehab2020} and \cite{Kilgard2020}, the rendering of a cusp with infinite curvature matches that of a near-cusp. The code to detect near-zero derivatives and re-evaluate is a small amount of logic that does not cause divergence issues, as opposed to extensive a separate ``regularization'' filter in \cite{Nehab2020} (Section 3.1 of the paper) or similar special-case handling.
```
Comment (armansito) [inline paper.tex]: > ...small amount of logic that does not cause divergence issues, as opposed to extensive a separate ``regularization'' filter in \cite{Nehab2020} (Section 3.1 of the paper) or similar special-case handling...

How about: ..._small amount of non-divergent logic, unlike the additional  ``regularization'' pass proposed in Section 3.1 of \\cite{Nehab2020}_ to handle special cases...

File: paper.tex [hunk 2]
```diff
@@ -413,6 +430,29 @@ \section{Performance}
 % TODO: discuss: GPUs tested, data sets. Test strategy
 % TODO: discuss: divergence (reference prim count)
 
+We measured the timing of our CPU implementation, generating both line and arc primitives, against Skia's stroker, which outputs quadratic Bézier primitives. The results are shown in Figure~\ref{fig:timings}. The time shown is the total for test files from the \emph{timings} dataset of \cite{Nehab2020}, and the tolerance in all cases was 0.25. TODO: measure against other implementations.
```
Comment (armansito) [inline paper.tex]: > ...which outputs quadratic Bézier primitives...

..._a mix of straight lines and quadratic Bézier primitives_...

File: paper.tex [hunk 3]
```diff
@@ -413,6 +430,29 @@ \section{Performance}
...
+
+As observed in \cite{Nehab2020}, Skia is the fastest stroke expansion algorithm by a considerable margin. The speed is attained with some compromises, in particular an imprecise error estimation that can sometimes generate outlines exceeding the given tolerance. We confirmed that this behavior is still present. The Nehab paper proposes a more careful error metric, but a significant fraction of the total computation time is dedicated to evaluating it.
```
Comment (armansito) [inline paper.tex]: > As observed in \cite{Nehab2020}, Skia...

This is mostly a nit but maybe rephrase this sentence to emphasize that this observation matches our own measurements, for example: _Similarly to \cite{Nehab2020}'s observations,_ Skia's stroke expansion algorithm is the fastest among the implementations we measured._

Comment (armansito) [inline paper.tex]: > We confirmed that this behavior is still present.

Add `\ref{fig:prim_count}` ?

**Review by armansito — APPROVED** [review]
Looks good, just a couple nits.

One thought I have about cusp_rendering.png: Should we make a version of this that has a smaller stroke width and the Euler segments are scaled up? The current appearance is accurate but the tiny cusp segment might be a little too tiny to see.

File: paper.tex (hunk 3 above)
Comment (raphlinus) [reply paper.tex]: prim_count is not evidence for this behavior. What would be is an example cubic that Skia handles incorrectly (https://gist.github.com/raphlinus/ec7722e8732cde7228fe61932c673b89), which results in `M0.564642 -0.825336Q92.728409 62.227280 170.805038 102.236145Q297.089844 166.948242 415.755798 190.532516Q567.199219 220.630859 702.725952 183.180008Q863.968750 138.621094 999.282654 -0.696646L1000.717346 0.696768Q600.928528 412.328979 -0.564642 0.825336L0.564642 -0.825336Z`. It's not clear to me whether we want to expend a figure on it, but perhaps we should, as it is a fairly important claim.
